File: android/app/src/main/java/io/quotawatch/ui/components/States.kt
package io.quotawatch.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Info
import androidx.compose.material.icons.outlined.Refresh
import androidx.compose.material.icons.outlined.Warning
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.quotawatch.data.Formatting
import io.quotawatch.net.ApiError
import io.quotawatch.ui.theme.QWColors
import java.time.Instant

/** 1dp separator — the open-layout divider (iOS Hairline). */
@Composable
fun Hairline(modifier: Modifier = Modifier) {
    Box(modifier.fillMaxWidth().height(1.dp).background(QWColors.Border))
}

/** Marks content as stale: 62% opacity (iOS qwStale). */
fun Modifier.qwStale(stale: Boolean): Modifier =
    if (stale) this.alpha(0.62f) else this

@Composable
fun LoadingStateView() {
    Column(
        modifier = Modifier.fillMaxWidth().padding(top = 120.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        CircularProgressIndicator(color = QWColors.Accent, strokeWidth = 2.dp, modifier = Modifier.size(28.dp))
        Text("正在读取最新配额", color = QWColors.Foreground, fontSize = 15.sp,
            modifier = Modifier.padding(top = 16.dp))
        Text("采集器通常会在几秒内回应", color = QWColors.Muted, fontSize = 13.sp,
            modifier = Modifier.padding(top = 6.dp))
    }
}

@Composable
fun ErrorStateView(error: ApiError?, onRetry: () -> Unit) {
    val isParse = error is ApiError.Decoding
    Column(
        modifier = Modifier.fillMaxWidth().padding(top = 120.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Icon(
            imageVector = if (isParse) Icons.Outlined.Info else Icons.Outlined.Warning,
            contentDescription = null,
            tint = QWColors.Muted,
            modifier = Modifier.size(32.dp),
        )
        Text(
            if (isParse) "这次数据没有读懂" else "无法连接到 Mac",
            color = QWColors.Foreground, fontSize = 16.sp,
            modifier = Modifier.padding(top = 14.dp),
        )
        Text(
            if (isParse) "采集器在线，但返回内容不完整。旧数据仍保留。"
            else error?.message ?: "检查 daemon 是否已用 --lan 启动，且手机与 Mac 在同一网络。",
            color = QWColors.Muted, fontSize = 13.sp, textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = 6.dp).padding(horizontal = 40.dp),
        )
        Row(modifier = Modifier.padding(top = 18.dp)) {
            androidx.compose.material3.TextButton(onClick = onRetry) {
                Icon(Icons.Outlined.Refresh, contentDescription = null,
                    tint = QWColors.Accent, modifier = Modifier.size(16.dp))
                Spacer(Modifier.size(6.dp))
                Text("重试连接", color = QWColors.Accent, fontSize = 14.sp)
            }
        }
    }
}

/** Offline banner: "离线 · 显示 Xs 前缓存". */
@Composable
fun StaleBanner(updatedAt: Instant?) {
    Row(
        modifier = Modifier.fillMaxWidth().padding(vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(Icons.Outlined.Warning, contentDescription = null,
            tint = QWColors.Warning, modifier = Modifier.size(14.dp))
        Spacer(Modifier.size(6.dp))
        Text(
            updatedAt?.let { "离线 · 显示 ${Formatting.ago(it)} 前缓存" } ?: "离线 · 显示最近缓存",
            color = QWColors.Warning, fontSize = 12.sp,
        )
    }
}

/** 44×44 circular utility button (iOS IconCircle). */
@Composable
fun IconCircle(onClick: () -> Unit, content: @Composable () -> Unit) {
    Box(
        modifier = Modifier
            .size(44.dp)
            .background(QWColors.Surface, CircleShape)
            .padding(0.dp),
        contentAlignment = Alignment.Center,
    ) {
        androidx.compose.material3.IconButton(onClick = onClick) { content() }
    }
}

File: ios/QuotaWatch/States.swift
import SwiftUI

// ─────────────────────────────────────────────────────────────
// 加载 / 错误 / 离线标识 —— 设计文档 03 · 全部状态。
// 文案说清发生了什么、还保留了什么，以及下一步在哪里；
// 刷新保留旧数据并降低透明度，不闪回骨架屏。
// ─────────────────────────────────────────────────────────────

/// 1px hairline —— 编辑版面建立秩序的主结构线
struct Hairline: View {
    var body: some View {
        Rectangle()
            .fill(QWColor.border)
            .frame(height: QWTokens.hairline)
    }
}

/// 首次读取加载态：说明在读什么、多久回应，不做骨架屏闪烁
struct LoadingStateView: View {
    var body: some View {
        VStack(spacing: 10) {
            ProgressView()
                .tint(QWColor.accent)
            Text("正在读取最新配额")
                .font(.system(size: 16, weight: .medium))
                .foregroundStyle(QWColor.foreground)
            Text("采集器通常会在几秒内回应")
                .font(.system(size: 13))
                .foregroundStyle(QWColor.subtle)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

/// 连接失败与解析失败是两个互斥变体，用具体文案区分
struct ErrorStateView: View {
    let error: APIError?
    let onRetry: () -> Void

    private var isParse: Bool { if case .decoding = error { return true }; return false }

    var body: some View {
        VStack(spacing: 8) {
            Image(systemName: isParse ? "doc.text.magnifyingglass" : "wifi.exclamationmark")
                .font(.system(size: 30, weight: .medium))
                .foregroundStyle(QWColor.warning)
            Text(isParse ? "这次数据没有读懂" : "无法连接到 Mac")
                .font(.system(size: 17, weight: .semibold))
                .foregroundStyle(QWColor.foreground)
            Text(detail)
                .font(.system(size: 14))
                .foregroundStyle(QWColor.subtle)
                .multilineTextAlignment(.center)
            Button(action: onRetry) {
                Text("重试连接")
                    .font(.system(size: 16, weight: .semibold))
                    .foregroundStyle(QWColor.accentInk)
                    .padding(.horizontal, 22)
                    .frame(height: 46)
                    .background(QWColor.accent,
                                in: RoundedRectangle(cornerRadius: QWTokens.Radius.control + 4, style: .continuous))
            }
            .buttonStyle(.plain)
            .padding(.top, 14)
        }
        .padding(.horizontal, QWTokens.Space.xl)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    private var detail: String {
        if isParse {
            return "采集器在线，但返回内容不完整。旧数据仍保留。"
        }
        return error?.errorDescription ?? "检查 daemon 是否已用 --lan 启动，且手机与 Mac 在同一网络。"
    }
}

/// 离线但有缓存 —— 数据继续可读，时效被明确标出
struct StaleBanner: View {
    let updatedAt: Date?

    var body: some View {
        HStack(spacing: 6) {
            Image(systemName: "wifi.slash")
                .font(.system(size: 11))
                .foregroundStyle(QWColor.subtle)
            Text(updatedAt.map { "离线 · 显示 \(Formatting.ago($0)) 前缓存" } ?? "离线 · 显示最近缓存")
                .font(.system(size: 13))
                .foregroundStyle(QWColor.subtle)
        }
        .padding(.vertical, 10)
    }
}
